tools/pubsys/src/aws/validate_ami/ami.rs

```diff
@@ -0,0 +1,195 @@
+//! The ami module owns the describing of images in EC2.
+
+use aws_sdk_ec2::model::Image;
+use aws_sdk_ec2::{Client as Ec2Client, Region};
+use futures::future::{join, ready};
+use futures::stream::{FuturesUnordered, StreamExt};
+use log::{info, trace};
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+use std::collections::HashMap;
+
+use crate::aws::ami::launch_permissions::{get_launch_permissions, LaunchPermissionDef};
+
+/// Structure of the EC2 image fields that should be validated
+#[derive(Serialize, Deserialize, PartialEq, Eq, Hash, Debug, Clone)]
+pub(crate) struct ImageDef {
+    /// The id of the EC2 image
+    pub(crate) id: String,
+
+    /// The name of the EC2 image
+    pub(crate) name: String,
+
+    /// Whether or not the EC2 image is public
+    #[serde(default)]
+    pub(crate) public: bool,
+
+    /// The launch permissions for the EC2 image.
+    pub(crate) launch_permissions: Option<Vec<LaunchPermissionDef>>,
+
+    /// Whether or not the EC2 image supports Elastic Network Adapter
+    #[serde(default = "default_ena_support")]
+    pub(crate) ena_support: bool,
+
+    /// The level of the EC2 image's Single Root I/O Virtualization support
+    #[serde(default = "default_sriov_net_support")]
+    pub(crate) sriov_net_support: String,
+}
+
+fn default_ena_support() -> bool {
+    true
+}
+
+fn default_sriov_net_support() -> String {
+    "simple".to_string()
+}
+
+impl From<Image> for ImageDef {
+    fn from(image: Image) -> Self {
+        Self {
+            id: image.image_id().unwrap_or_default().to_string(),
+            name: image.name().unwrap_or_default().to_string(),
+            public: image.public().unwrap_or_default(),
+            // Will be populated later if the image is not expected to be public. If it is expected
+            // to be public, then individual launch permissions are irrelevant
+            launch_permissions: None,
+            ena_support: image.ena_support().unwrap_or_default(),
+            sriov_net_support: image.sriov_net_support().unwrap_or_default().to_string(),
+        }
+    }
+}
+
+pub(crate) async fn describe_images<'a>(
+    clients: &'a HashMap<Region, Ec2Client>,
+    image_ids: &HashMap<Region, Vec<String>>,
+    expected_image_public: &HashMap<String, bool>,
+) -> HashMap<&'a Region, Result<HashMap<String, ImageDef>>> {
+    // Build requests for images; we have to request with a regional client so we split them by
+    // region
+    let mut requests = Vec::with_capacity(clients.len());
+    for region in clients.keys() {
+        trace!("Requesting images in {}", region);
+        let ec2_client: &Ec2Client = &clients[region];
```

Comment:
```suggestion
        let ec2_client: &Ec2Client = &clients.get(region).context()...;
```

Indexing with `[]` can panic.